src/include/daos_srv/srv_csum.h

```diff
@@ -50,4 +31,165 @@ ds_csum_add2iod(daos_iod_t *iod, struct daos_csummer *csummer,
+	SCRUB_STATUS_UNKNOWN = 0,
+	SCRUB_STATUS_RUNNING = 1,
+	SCRUB_STATUS_NOT_RUNNING = 2,
+};
+
+/*
+ * telemetry metrics for the scrubber
+ * The target must be parsed out by
+ * lib/telemetry/promexp/collector.go:extractLabels(). If the format here
+ * changes, extractLabels() must be updated as well.
+ */
+#define DF_POOL_DIR "%s/target:%d/scrubber"
+#define DP_POOL_DIR(ctx) (ctx)->sc_pool->sp_path, \
+			(ctx)->sc_dmi->dmi_tgt_id
+
+#define M_CSUM_COUNTER "csums/current"
+#define M_CSUM_TOTAL_COUNTER "csums/total"
+#define M_CSUM_PREV_COUNTER "csums/prev"
+#define M_CSUM_CORRUPTION "corruption/current"
+#define M_CSUM_TOTAL_CORRUPTION "corruption/total"
+#define M_STARTED "scrubber_started"
+#define M_LAST_DURATION "last_duration"
+
+#define m_inc_counter(m) d_tm_inc_counter((m), 1)
+#define m_reset_counter(m) d_tm_set_counter((m), 0)
+
+struct scrub_metrics {
+	struct d_tm_node_t *sm_start;
+	struct d_tm_node_t *sm_last_duration;
+	struct d_tm_node_t *sm_csum_calcs;
+	struct d_tm_node_t *sm_total_csum_calcs;
+	struct d_tm_node_t *sm_last_csum_calcs;
+	struct d_tm_node_t *sm_corruption;
+	struct d_tm_node_t *sm_total_corruption;
+};
+
+struct scrub_ctx_metrics {
+	struct d_tm_node_t	*scm_pool_ult_start;
+	struct d_tm_node_t	*scm_pool_ult_wait_time;
+	struct scrub_metrics	 scm_pool_metrics;
+};
+
+/* Scrub the pool */
+struct scrub_ctx {
+	/**
+	 * Metrics gathered during scrubbing process
+	 */
+	struct scrub_ctx_metrics sc_metrics;
+
+	struct dss_module_info	*sc_dmi;
+
+	/**
+	 * Pool
+	 **/
+	uuid_t			 sc_pool_uuid;
+	daos_handle_t		 sc_vos_pool_hdl;
+	struct ds_pool		*sc_pool; /* Used to get properties */
+	struct timespec		 sc_pool_start_scrub;
+	int			 sc_pool_last_csum_calcs;
+	int			 sc_pool_csum_calcs;
+
+	/**
+	 * Container
+	 **/
+	/* callback function that will provide the csummer for the container */
+	ds_get_cont_fn_t	 sc_cont_lookup_fn;
+	ds_put_cont_fn_t	 sc_cont_put_fn;
+	ds_cont_is_stopping_fn_t sc_cont_is_stopping_fn;
+	struct cont_scrub	 sc_cont;
+
+	/** Number of msec between checksum calculations */
+	daos_size_t		 sc_msec_between_calcs;
+
+	/**
+	 * Object
+	 */
+	daos_unit_oid_t		 sc_cur_oid;
+	daos_key_t		 sc_dkey;
+	struct dcs_csum_info	*sc_csum_to_verify;
+	daos_epoch_t		 sc_epoch;
+	uint16_t		 sc_minor_epoch;
+	daos_iod_t		 sc_iod;
+
+	/* Current vos object iterator */
+	daos_handle_t		 sc_vos_iter_handle;
+
+	/* Schedule controlling function pointers and arg */
+	uint32_t		 sc_credits_left;
+	ds_sleep_fn_t		 sc_sleep_fn;
+	ds_yield_fn_t		 sc_yield_fn;
+	void			*sc_sched_arg;
+
+	enum scrub_status	 sc_status;
+	bool			 sc_did_yield;
+};
+
+/*
+ * It is expected that the pool uuid/handle and any functional dependencies are
+ * set in the scrubbing context. The container/object info should not be set.
+ * This function will iterate over all of the containers in the pool and if
+ * checksums are enabled on the pool, each object in the container will be
+ * scrubbed.
+ */
+int ds_scrub_pool(struct scrub_ctx *ctx);
```

Comment:
After a conversation with Niu, we determined to leave it like this for now because internal vos api is needed to get the bio context that bio_read() needs. Will look in the future at possible exposing more from VOS so external can do a bio_read() and then might move vos_pool_scrub back to srv_pool_scrub. Will update the ds_scrub_pool to be a VOS api for now.